### ISSUE TITLE
Rewriting and universes

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#2090]
The code below should typecheck but it doesn’t with the error

```
HetEq idp u v != u == v of type Set (.j .Agda.Primitive.⊔ .i)
when checking that the expression idr has type
HetEq idp u v ↦ (u == v)
```

Note that it does work if `A` and `B` (in the last definition) are both in the same universe `Set i`, so it seems to be a problem with the interaction with the universe levels somehow.

``` agda
{-# OPTIONS --rewriting #-}

postulate
  _↦_ : ∀ {i} {A : Set i} → A → A → Set i
  idr : ∀ {i} {A : Set i} {a : A} → a ↦ a
{-# BUILTIN REWRITE _↦_ #-}

postulate
  _==_ : ∀ {i} {A : Set i} → A → A → Set i
  idp : ∀ {i} {A : Set i} {a : A} → a == a

postulate
  HetEq : ∀ {i} {A B : Set i} (e : A == B) (a : A) (b : B) → Set i
  HetEq↦ : ∀ {i} {A : Set i} (a b : A) → HetEq idp a b ↦ (a == b)
  {-# REWRITE HetEq↦ #-}

HetEq↦2 : ∀ {i j} {A : Set i} {B : Set j} (u v : A → B) → HetEq idp u v ↦ (u == v)
HetEq↦2 u v = idr
```

**Comment by jespercockx** [comment]
The problem seems to be that the matching algorithm doesn't know about the additional (in)equalities between universe levels, in this case that `lsuc i ⊔ lsuc j == lsuc (i ⊔ j)`. It really should take these equalities into account, and others as well such as `i ⊔ j == j ⊔ i` etc. Now we only need to find a principled way to do matching modulo these equations...

**Comment by jespercockx** [comment]
The problem doesn't seem to be solvable in general: how do you match something of the form `(n1 + t1) ⊔ (n2 + t2) ⊔ ... ⊔ (nk + tk)` against a pattern of the form `(m1 + p1) ⊔ ... ⊔ (ml + pl)`? There's no way to decide which part of the pattern to match with which part of the term. However, it's possible to do something simpler that only works for `lsuc` but would at least fix this specific example. Working on it...

**Comment by asr** [comment]
@jespercockx since you closed the issue, please add the 2.5.2 milestone.

**Comment by asr** [comment]
Thanks!
